Add channel class and task method for parallel processing in mruby

naritta opened the pull request:
This PR is for https://github.com/h2o/h2o/issues/1329

TODO:
- [x]  add queue for when ch.push is called before ch.shift
- [x]  add tests
- [x]  use script to create embedded code

kazuho commented on include/h2o/mruby_.h:
```diff
@@ -177,10 +182,18 @@ mrb_value h2o_mruby_http_join_response_callback(h2o_mruby_context_t *ctx, mrb_va
 mrb_value h2o_mruby_http_fetch_chunk_callback(h2o_mruby_context_t *ctx, mrb_value receiver, mrb_value input,
                                               int *next_action);
+mrb_value h2o_mruby_channel_shift_callback(h2o_mruby_context_t *ctx, mrb_value receiver, mrb_value input,
+                                              int *next_action);
```
How about retaining the queue in mruby side, while restricting use of the C code to when signalling between fibers is necessary?

I believe that doing so would simplify the code and will give us more flexibility in extending the object (for example, it would be easier to add a `peek` method). 

```ruby
class Channel
  def initialize()
    @queue = []
  end
  def push(o)
    @queue << o
    self._notify
  end
  def shift
    while true
      if !@queue.empty?
        return @queue.shift
      end
      self._wait
    end
  end
  // _notify and _wait are C methods that are called via fiber switching
end
```

naritta replied:
I agree this idea, because it's a little complex to deal with queue in C.
Then I fixed it like below. (I'll add rb file and use script to create embedded code soon.)
https://github.com/h2o/h2o/pull/1336/commits/064a44b6bff761154b224e4b3c253553d4d69fd8

・For ```_h2o__channel_wait```, it seems we can't call fiber yield ruby code from C, then I wrote to call ```_h2o__channel_wait``` directly.
・For initializing Channel object, maybe there is better or more simple way, then I'm willing to fix it.

naritta commented on lib/handler/mruby/channel.c:
```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2015-2016 DeNA Co., Ltd., Kazuho Oku
```
I appreciate your kindness.
I fixed it here: https://github.com/h2o/h2o/pull/1336/commits/60f1d0eaf56ed62a477a1a5be85f70e25798ea3d

naritta commented:
@kazuho Thank you for the comments. I'm really grateful to be involved with such great product.

I added task method, and it works like below.
・When task with block is called, block is processed firstly.
・In the block, when async function is called, the fiber will be yielded in this method, and pass the resumer and object to context.
・When event is called, it's called from mruby_run_fiber as usual and it returns output into mruby_run_fiber. When the fiber ends, it returns 0 to mruby_run_fiber method and finish running.

Perhaps it can be improved, then please let me know when there is something. I'm willing to fix it.

naritta commented:
@kazuho @i110 
I've implemented channel class and task method.
Could you review it when you have time?

i110 commented:
@naritta sorry for being late. I'll review this on Wednesday

naritta commented:
@i110 @kazuho 
Sorry to bother though you are so busy, but I'm grateful if you could review this because I want to use h2o with this function in my company product.

tagomoris commented:
ping?

i110 commented:
@naritta @tagomoris Sorry for my long absence. I reviewed and added some inline comments.
By the way, when the discussion gets complicated, please feel free to use japanese to clalify the discussion!

naritta commented:
@i110 I can't see your review comments anywhere, Where can I see it?

i110 commented:
@naritta Sorry I forgot to submit reviews.

naritta commented:
Thank you for reviewing, I'll fix it.

naritta commented on lib/handler/mruby/http_request.c:
```diff
@@ -320,6 +320,21 @@ static int flatten_request_header(h2o_mruby_shared_context_t *shared_ctx, h2o_io
+    struct st_h2o_mruby_http_request_context_t *ctx;
+
+    mrb_value receiver;
+    mrb_value self_obj;
```
I had mistakes, I fixed it to use self.
https://github.com/h2o/h2o/pull/1336/commits/6316078f8bc9f4d7ece41ae1d7f43e631ff9b0c2

naritta commented on include/h2o/mruby_.h:
```diff
@@ -177,10 +182,18 @@ mrb_value h2o_mruby_http_join_response_callback(h2o_mruby_context_t *ctx, mrb_va
+/* handler/mruby/channel.c */
+void h2o_mruby_channel_init_context(h2o_mruby_shared_context_t *ctx);
+
+mrb_value h2o_mruby_channel_shift_callback(h2o_mruby_context_t *ctx, mrb_value receiver, mrb_value input,
```
I deleted duplicated definition.
https://github.com/h2o/h2o/pull/1336/commits/792f3f15e0310c6b42399a73b8c3541c07883f74

naritta commented on lib/handler/mruby/embedded/core.rb:
```diff
@@ -25,6 +25,24 @@ def _h2o_eval_conf(__h2o_conf)
+      block.call
+      # For when it's called in h2o_mruby_run_fiber and return output,
+      # or block doesn't have asynchronous callback
+      Fiber.yield([0, nil, nil])
```
I fixed to  simply returns `[0, nil, nil]`
https://github.com/h2o/h2o/pull/1336/commits/ea711f5b97b4258956e6213e01ae1df84e2f2bfd

naritta commented on lib/handler/mruby/channel.c:
```diff
@@ -0,0 +1,143 @@
+    return mrb_nil_value();
+}
+
+static mrb_value create_channel_method(mrb_state *mrb, mrb_value self)
```
I fixed to define normal initialize method for initializing channel.
https://github.com/h2o/h2o/pull/1336/commits/4c17ebd171223143bd5f8b7cc254f6229ea86d6b

naritta commented on lib/handler/mruby/embedded/channel.rb:
```diff
@@ -0,0 +1,39 @@
+        if !@queue.empty?
+          return @queue.shift
+        end
+        _h2o__channel_wait(self)
```
I fixed not to use needless loop.
https://github.com/h2o/h2o/pull/1336/commits/82fa91dfd50ee66ad90905a6097683ce01fe0878

naritta commented on lib/handler/mruby/channel.c:
```diff
@@ -0,0 +1,143 @@
+        h2o_mruby_run_fiber(ctx->ctx, detach_receiver(ctx), mrb_nil_value(), NULL);
+        h2o_mruby_shared_context_t *shared_ctx = mrb->ud;
+        /* When it's called in task, retrieve current_context for next action in task */
+        shared_ctx->current_context = ctx->ctx;
```
When creating task like this:
`task {
     http_request(req_url).join;
     ch.push "hello";
     http_request(req_url).join;
   }`

channel_notify_method calls h2o_mruby_run_fiber, and sets current_context NULL in this h2o_mruby_run_fiber.
And after h2o_mruby_run_fiber, it continues processing task block, but current_context is NULL, then http_request_method cannot get current_context.

I think another way is not setting current_context NULL if it's call by channel_notify_method.
If there are other better way, please tell me. I'll fix it.

i110 replied:
The most straightforward solution might be doing like this in `h2o_mruby_run_fiber`:
```
h2o_mruby_context_t *old = ctx->shared->current_context;
ctx->shared->current_context = ctx;
/* ... */
ctx->shared->current_context = old;
```

But in my thoughts, it's better to eliminate nested h2o_mruby_run_fiber call. How about using `yield` to switch fibers?

naritta replied:
I tried to use `yield ` to switch fibers and use the run loop in h2o_mruby_run_fiber, but it says
`mruby raised: (eval):5: can't cross C function boundary (FiberError)`
It seems impossible to yield from mruby->C->mruby.
If my understanding about this is not correct, please tell me.
Or I'll fix to use straightforward solution above.

naritta commented on lib/handler/mruby/embedded/core.rb:
```diff
@@ -25,6 +25,24 @@ def _h2o_eval_conf(__h2o_conf)
+    klass = fiber_res[2][0]
+    # This should be called only one time.
+    # After that, the fiber is called in mruby_run_fiber and it register receiver in it.
+    klass.register_receiver(receiver, klass)
```
>What does klass means? For example, in channel usage, fiber_res[2][0] seems to be H2O::Channel 's instance, not class.

I think we use `klass` for instance name in ruby usage. If others is better, I'll fix it.
https://stackoverflow.com/questions/4299289/what-is-the-difference-between-class-and-klass-in-ruby

naritta replied:
>Is it guaranteed that the klass object respond_to? :register_receiver ? I mean that many asynchronous functions can be called in task block, and HTTP response body streaming is one of the examples. So if I understand correctly, http_request(..).join[2].join fails (but sorry I haven't tried)

It's not guaranteed and we must create register_receiver when we create new class having async function. (I forgot to add for HttpInputStream, then I added here.)
https://github.com/h2o/h2o/pull/1336/commits/aa53e9023d33d26061c2674012b2a0c089e1964e

naritta replied:
>When the block yileds (calls defined callback method), there may be some args (i.e. fiber_res[2][1], fiber_res[2][2]..) How to handle that?

sorry, I realized `h2o_mruby_http_fetch_chunk_callback` are using args other than self in new code now.
How about this idea?:
1. move parts using args to new function named reserved word like `process_args(args)`
2. If there are some value on args other than self(index 0) in async function, call `process_args(args)`
3. do same thing for task function calling `process_args(args)`

For example, In case of `_h2o__http_fetch_chunk`,
1. move this(https://github.com/h2o/h2o/blob/master/lib/handler/mruby/http_request.c#L463-L471) from `h2o_mruby_http_fetch_chunk_callback` to new function `process_args`, and call `process_args`
2. In task function, if there is common same named function `process_args`.

In this way, we don't have to create branch in task function like mruby_run_fiber because of using same named function, but we have to care when we add function using args.

naritta commented:
@i110 Thank you for reviewing and sorry to be late, I fixed them.
Could you review it when you have time?

i110 commented:
@naritta Thank you for working on this.

Regarding the current design, I feel that it's ちょっとだるい to define `register_receiver` function for every asynchronous features (like redis, sleep, etc..), but otherwise it can't be used in `task` method. Also, the name of `register_receiver` is some kind of implicit conventions that the implementors must be aware of.

How about doing like this? I think this is much simpler than the current implementation, because all of the things required for `task` feature can be done by this small patch, without introducing any new C functions and conventions of that!

```
diff --git a/include/h2o/mruby_.h b/include/h2o/mruby_.h
index 37fa4ada..d045faef 100644
--- a/include/h2o/mruby_.h
+++ b/include/h2o/mruby_.h
@@ -128,6 +128,7 @@ typedef struct st_h2o_mruby_generator_t {
 #define H2O_MRUBY_CALLBACK_ID_HTTP_FETCH_CHUNK -6
 #define H2O_MRUBY_CALLBACK_ID_REDIS_JOIN_REPLY -7
 #define H2O_MRUBY_CALLBACK_ID_SLEEP -999
+#define H2O_MRUBY_CALLBACK_ID_RUN_CHILD_FIBER -777

 #define h2o_mruby_assert(mrb)                                                                                                      \
     if (mrb->exc != NULL)                                                                                                          \
diff --git a/lib/handler/mruby.c b/lib/handler/mruby.c
index 0f3361b6..003e35e9 100644
--- a/lib/handler/mruby.c
+++ b/lib/handler/mruby.c
@@ -759,6 +759,8 @@ void h2o_mruby_run_fiber(h2o_mruby_context_t *ctx, mrb_value receiver, mrb_value
     mrb_int status;
     h2o_mruby_generator_t *generator = NULL;

+    mrb_value resumers = mrb_ary_new(mrb); /* TODO this array should be re-used, protected from gc, and asserted being empty here*/
+
     while (1) {
         /* send input to fiber */
         output = mrb_funcall_argv(mrb, receiver, ctx->shared->symbols.sym_call, 1, &input);
@@ -775,7 +777,7 @@ void h2o_mruby_run_fiber(h2o_mruby_context_t *ctx, mrb_value receiver, mrb_value
         status = mrb_fixnum(v);

         /* if no special actions were necessary, then the output is a rack response */
-        if (status >= 0)
+        if (status > 0)
             break;

         /* take special action depending on the status code */
@@ -832,13 +834,24 @@ void h2o_mruby_run_fiber(h2o_mruby_context_t *ctx, mrb_value receiver, mrb_value
             case H2O_MRUBY_CALLBACK_ID_SLEEP:
                 input = h2o_mruby_sleep_callback(ctx, receiver, args, &run_again);
                 break;
+            case H2O_MRUBY_CALLBACK_ID_RUN_CHILD_FIBER: {
+                mrb_value resumer = mrb_ary_entry(args, 0);
+                mrb_ary_push(mrb, resumers, resumer);
+                input = mrb_nil_value();
+                run_again = 1;
+                break;
+            }
             default:
                 input = mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "unexpected callback id sent from rack app");
                 run_again = 1;
                 break;
             }
-            if (run_again == 0)
-                goto Exit;
+            if (run_again == 0) {
+                if (RARRAY_LEN(resumers) == 0)
+                    goto Exit;
+                input = mrb_nil_value();
+                receiver = mrb_ary_pop(mrb, resumers);
+            }

         } else {
             input = mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "callback from rack app did not receive an array arg");
diff --git a/lib/handler/mruby/embedded/core.rb b/lib/handler/mruby/embedded/core.rb
index e0f3a5c7..17655caf 100644
--- a/lib/handler/mruby/embedded/core.rb
+++ b/lib/handler/mruby/embedded/core.rb
@@ -29,6 +29,15 @@ module Kernel
   H2O_CALLBACK_ID_EXCEPTION_RAISED = -1
   H2O_CALLBACK_ID_CONFIGURING_APP  = -2
   H2O_CALLBACK_ID_CONFIGURED_APP   = -3
+  H2O_CALLBACK_ID_RUN_CHILD_FIBER  = -777
+
+  def task(&block)
+    fiber = Fiber.new {
+      block.call
+      [0]
+    }
+    Fiber.yield([H2O_CALLBACK_ID_RUN_CHILD_FIBER, proc { fiber.resume }, [_h2o_create_resumer()]])
+  end

   def _h2o_define_callback(name, id)
     Kernel.define_method(name) do |*args|
```

for example:

```ruby
        mruby.handler: |
          puts "# before defining task.."
          task {
            3.times {|i|
              puts "# before http_request (#{i}).."
              status, headers, body = http_request('http://example.com/').join
              puts "# after http_request (#{i}) (status + #{status})"
            }
          }
          puts "# after defining task, will finish configuration phase!"

          proc {|env|
            [200, {}, ["hello\n"]]
          }
```

outputs the following.
```
# before defining task..
# before http_request (0)..
# after defining task, will finish configuration phase!
# after http_request (0) (status + 200)
# before http_request (1)..
# after http_request (1) (status + 200)
# before http_request (2)..
# after http_request (2) (status + 200)
```

But sorry I had no time to verify my idea and impl, so I would appreciate if you could verify and make it trusty. Of cause before that, please let me know your idea on this 🍣

i110 commented on lib/handler/mruby/channel.c:
```diff
@@ -0,0 +1,141 @@
+    h2o_mruby_shared_context_t *shared_ctx = mrb->ud;
+
+    struct st_h2o_mruby_channel_context_t *ctx;
+    ctx = h2o_mem_alloc(sizeof(*ctx));
```
This seems to leak. You have to free this in on_gc_dispose_channel

naritta replied:
I fixed to free ctx and check null of values of that.
https://github.com/h2o/h2o/pull/1336/commits/05790cd65a528011ea1702d33ec75669e7b97915

i110 commented on include/h2o/mruby_.h:
```diff
@@ -121,6 +125,7 @@ typedef struct st_h2o_mruby_generator_t {
 #define H2O_MRUBY_CALLBACK_ID_SEND_CHUNKED_EOS -4
 #define H2O_MRUBY_CALLBACK_ID_HTTP_JOIN_RESPONSE -5
 #define H2O_MRUBY_CALLBACK_ID_HTTP_FETCH_CHUNK -6
+#define H2O_MRUBY_CALLBACK_ID_CHANNEL_SHIFT -7
```
`H2O_MRUBY_CALLBACK_ID_CHANNEL_WAIT` might be better name?

naritta replied:
I fixed to use name `H2O_MRUBY_CALLBACK_ID_CHANNEL_WAIT`
https://github.com/h2o/h2o/pull/1336/commits/36f718e3c2ac4acde3a48c8057b132d0293531f3

i110 commented on lib/handler/mruby/channel.c:
```diff
@@ -0,0 +1,141 @@
+static mrb_value channel_notify_method(mrb_state *mrb, mrb_value self)
+{
+    struct st_h2o_mruby_channel_context_t *ctx;
+    ctx = DATA_PTR(self);
```
it's better to use `mrb_data_check_get_ptr` instead of `DATA_PTR` macro, or at least be consistent

i110 commented on t/50mruby-channel.t:
```diff
@@ -0,0 +1,95 @@
+    is $stdout, 1;
+};
+
+plan skip_all => 'mruby support is off'
```
JFYI, you don't have to call `skip_all` multiple times

naritta replied:
I deleted needless `skip_all `
https://github.com/h2o/h2o/pull/1336/commits/8f4c9bf1e33ec15db341eccd52b94ead21f32267

i110 commented on t/50mruby-channel.t:
```diff
@@ -0,0 +1,95 @@
+      /:
+        mruby.handler: |
+          Proc.new do |env|
+            ch = create_channel
```
H2O::Channel.new?

naritta replied:
I fix to use new for creating instances.
https://github.com/h2o/h2o/pull/1336/commits/7edafbc7cb5f5e1960075a59a4bb30f2b5efe5b6

i110 commented on t/50mruby-task.t:
```diff
@@ -0,0 +1,123 @@
+    is $stdout, "123456";
+};
+
+done_testing();
```
I would appreciate if you could add the tests which joins response body :D

naritta replied:
I added a test for that.
https://github.com/h2o/h2o/pull/1336/commits/250acdd7c8e59b84b3460db4b6c1d7ff9bf63842

naritta commented on lib/handler/mruby/embedded/channel.rb:
```diff
@@ -0,0 +1,39 @@
+      if @queue.empty?
+        _h2o__channel_wait(self)
+      end
+      begin
```
I deleted begin ~ rescue
https://github.com/h2o/h2o/pull/1336/commits/7a0521597df00417e6a25f462ab23018c24f01e2

naritta commented:
Thank you for sharing great idea about deleting `register_receiver`. I could make sure it works well.
Then I fixed to use this and deleted `register_receiver`.
https://github.com/h2o/h2o/pull/1336/commits/25373b4a8412c5cdeec65128e938e0e7ba093411

And also, I fixed other points. Could you review them when you have time?

naritta commented:
@i110 I applied clang-format and fixed failed CI.
Could you review this PR?

tagomoris commented:
@i110 @kazuho Could you review this change if possible? We want to implement many application features on this change 🙏

i110 commented:
@naritta @tagomoris Sorry for my belated response.. I filed https://github.com/naritta/h2o/pull/1 to your forked repo, which merges the latest master and contains some fixes. If it's merged, the changes will appear here :D

naritta commented:
@i110 Thank you for reviewing and fixing. I could make sure it works well.

i110 commented on lib/handler/mruby.c:
```diff
@@ -844,17 +873,24 @@ void h2o_mruby_run_fiber(h2o_mruby_context_t *ctx, mrb_value receiver, mrb_value
         mrb_gc_protect(mrb, input);
     }
 
+    if (status == 0) {
```
Sorry I forgot to remove this check and goto. I introduced _h2o__finish_child_fiber, so this is no longer needed. Could you remove this?

i110 commented:
From my viewpoint this PR seems to be ready to merged. @kazuho Any thoughts?

kazuho commented on lib/handler/mruby/embedded.c.h:
```diff
@@ -79,6 +79,13 @@
+    "  def task(&block)\n"                                                                                                         \
+    "    fiber = Fiber.new do\n"                                                                                                   \
+    "      block.call\n"                                                                                                           \
+    "      _h2o__finish_child_fiber()\n"                                                                                           \
```
Would you mind elaborating what happens if `block.call` raises an exception? My expectation is that `h2o__finish_child_fiber()` will not be called in such case.

If calling `h2o__finish_child_fiber()` makes any difference, then I believe that we should call the function even if `block.call` raises an error. Or if `h2o__finish_child_fiber()` is a noop, can we eliminate the function?

naritta replied:
For now, entire application will stop when `block.call` raises.
Then I fixed to call `_h2o__send_error` and not to call `_h2o__finish_child_fiber` intentionally.
By this, task will raise an error and only task dies but parent application will not stop when fiber in task raises.
If I should make it parent application stop, I'll fix it.
https://github.com/h2o/h2o/pull/1336/commits/52ea22320fa0c178705b4b041495dcb4f3f85dde

i110 replied:
perfect! :D

kazuho commented on lib/handler/mruby/channel.c:
```diff
@@ -0,0 +1,126 @@
+
+static void attach_receiver(struct st_h2o_mruby_channel_context_t *ctx, mrb_value receiver)
+{
+    assert(mrb_nil_p(ctx->receiver));
```
Could you please elaborate how we ensure that a Channel will have at most one listeners?

I couldn't find code that provides such guarantee, and wonder what happens if two fibers call `shift` of one channel object.

kazuho replied:
PS. Please note that I am not suggesting that we should support multiple readers. Supporting only one reader is fine, but an error in mruby script should be reported as an mruby exception; not an abrupt termination of the server process.

naritta replied:
I fixed to raise error when shift is called though it's already registered.
https://github.com/h2o/h2o/pull/1336/commits/2cf40c2ee4c2c78dcef25c245ee9ad1a2004c9c9

i110 replied:
@naritta Would you mind supporting multiple listener (shifter)? Discussed with @kazuho today, it may be useful in minor cases and can be easily implemented .

To implement that, simply do:
- change receiver to receivers (Array)
- use mrb_ary_push instead of attach_receiver
- use mrb_ary_shift instead of detach_receiver

How do you feel?

naritta replied:
Sounds cool. I fixed to support multiple listener like above.
https://github.com/h2o/h2o/pull/1336/commits/916cab15f8f88cfb608b769379f6af4139fb8612

kazuho commented on lib/handler/mruby/channel.c:
```diff
@@ -0,0 +1,126 @@
+{
+    struct st_h2o_mruby_channel_context_t *ctx = _ctx;
+    assert(ctx != NULL); /* ctx can only be disposed by gc, so data binding has been never removed */
+    free(ctx);
```
Do we need to call `mrb_gc_unregister(mrb, ctx->receiver)`?

I believe that the question boils down to if there is a chance in GC collecting the channel object while a receiver is being attached to a object; and to me, something like below seems to trigger such condition:
```
task {
  H2O::Channel.new.shift
}
```

naritta replied:
I fixed to call mrb_gc_unregister.
https://github.com/h2o/h2o/pull/1336/commits/da33b132c18034b76d9965cee68469b2ff1767f8

naritta commented:
@kazuho Thank you for reviewing. I fixed them and I commented. Could you check it?

i110 commented on lib/handler/mruby/channel.c:
```diff
@@ -0,0 +1,131 @@
+{
+    struct st_h2o_mruby_channel_context_t *ctx = _ctx;
+    assert(ctx != NULL); /* ctx can only be disposed by gc, so data binding has been never removed */
+    mrb_gc_unregister(mrb, ctx->receiver);
```
Could you add a check that ensures ctx->receiver is not nil? i.e. if (!mrb_nil_p(ctx->receiver))

naritta replied:
I fixed to check if `ctx->receiver` is not nil before `mrb_gc_unregister `.
https://github.com/h2o/h2o/pull/1336/commits/4312b539294bbb81d7cdf1621a05a08e53c2e0dd

i110 replied:
At this time we definitely set `receivers` in initialization, so this nil check you newly added is no longer needed. Sorry for the confusion 😵

naritta replied:
Sorry, I forgot to check it. I deleted it.
https://github.com/h2o/h2o/pull/1336/commits/03c0a42d76750c3cd95f1eab894be9151ac53b5e

naritta commented:
@i110  Thank you for reviewing. I fixed to support multiple listener and could you check it?

i110 commented on lib/handler/mruby/channel.c:
```diff
@@ -73,7 +56,7 @@ static mrb_value channel_initialize_method(mrb_state *mrb, mrb_value self)
     assert(shared_ctx->current_context != NULL);
     ctx->ctx = shared_ctx->current_context;
-    ctx->receiver = mrb_nil_value();
+    ctx->receivers = mrb_ary_new(mrb);
```
`receivers` should be `mrb_gc_register` ed?

naritta replied:
Thanks, I see. Do you think it should be `mrb_gc_register ` or `mrb_gc_protect` after shited also?

i110 replied:
Sorry I don't understand your question. mrb_gc_registerとmrb_gc_protectのどちらを使うべきかという意味ですか？であれば、ここではregisterのほうですね。 (protectだとgc arena indexがrestoreされたらgcされてしまうので）
それともshiftされたobjectも保護すべきかどうか、という質問ですかね？ (shited => shifted?)

naritta replied:
Sorry for confusing.
shiftされたobjectも保護すべきかどうかという質問です。

naritta replied:
Is it enough because mrb_gc_arena_save is done here?
https://github.com/naritta/h2o/blob/916cab15f8f88cfb608b769379f6af4139fb8612/lib/handler/mruby/channel.c#L74-L77

i110 replied:
Ah, I think you have to call `mrb_gc_protect(mrb, shifted_receiver)` immediately after shifting from receivers. Otherwise it can be GCed in mrb_funcall function before `call` method of that Proc is called.

naritta replied:
Thanks, I see. I fixed to call `mrb_gc_protect ` here.
https://github.com/h2o/h2o/pull/1336/commits/8926f7d74423c852f6ab10e58fda90e0fda6aee6

i110 commented:
Thank you so much for your effort!

naritta commented:
Thank you for review and great opportunity!
